File: vol2/ch23/LeetCode1306-1.cpp
class Solution {
    vector<bool> visited;
public:
    bool canReach(vector<int>& arr,int start) {
        int n=arr.size();
        visited=vector<bool>(n,false);
        return dfs(arr,start);
    }
    bool dfs(vector<int>& arr,int i) {
        if (i<0 || i>=arr.size())
            return false;
        if(visited[i]) return false;
        if(arr[i]==0) return true;
        visited[i]=true;
        if(dfs(arr,i+arr[i])) return true;
        if(dfs(arr,i-arr[i])) return true;
        return false;
    }
};

File: vol2/ch23/LeetCode1306-1.py
class Solution:
    def canReach(self,arr:List[int],start:int)->bool:
        n=len(arr)
        self.visited=[False for _ in range(n)]
        return self.dfs(arr,start)
    def dfs(self,arr,i):
        if i<0 or i>=len(arr):
            return False
        if self.visited[i]:return False
        if arr[i]==0:return True
        self.visited[i]=True
        if self.dfs(arr,i+arr[i]):return True
        if self.dfs(arr,i-arr[i]): return True
        return False
